transaction/core/src/ring_ct/signing_digest.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{
+    domain_separators::{EXTENDED_MESSAGE_AND_TX_SUMMARY_DOMAIN_TAG, EXTENDED_MESSAGE_DOMAIN_TAG},
+    tx::TxPrefix,
+    BlockVersion, CompressedCommitment, TxSummary,
+};
+use alloc::vec::Vec;
+use mc_crypto_digestible::{DigestTranscript, Digestible, MerlinTranscript};
+use mc_util_zip_exact::ZipExactError;
+
+/// The MLSAG signing digest is the digest that MLSAGs actually sign
+pub struct MLSAGSigningDigest(pub Vec<u8>);
+
+/// The extended message digest (or, before block version 2, the extended
+/// message)
+pub struct ExtendedMessageDigest(pub Vec<u8>);
+
+impl From<MLSAGSigningDigest> for Vec<u8> {
+    fn from(src: MLSAGSigningDigest) -> Self {
+        src.0
+    }
+}
+
+impl From<ExtendedMessageDigest> for Vec<u8> {
+    fn from(src: ExtendedMessageDigest) -> Self {
+        src.0
+    }
+}
+
+/// Compute the digest that mlsags should actually sign, depending on the block
+/// version.
+///
+/// Returns:
+/// * The MLSAG Signing Digest
+/// * The TxSummary
+/// * The extended_message_digest.
+///
+/// TODO: When support for block version < 2 is deprecated,
+/// we can make this return `[u8; 32]` instead of Vec<u8>, which is nicer for
+/// the hardware wallet implementation.
+pub fn compute_mlsag_signing_digest(
+    block_version: BlockVersion,
+    tx_prefix: &TxPrefix,
+    pseudo_output_commitments: &[CompressedCommitment],
+    range_proof_bytes: &[u8],
+    range_proofs: &[Vec<u8>],
```

Comment:
There's some docu in the other file, probably should document this one also though.

The issue is like:

In block version 2, there is only one range proof that covers all of the pseudo-output and output commitments. This works because in block version 2, there are no mixed transactions, so they all have the same token id and the same Pedersen generators. (in this version, range_proofs is empty)

In block version 3, mixed transactions are allowed, but the bulletproofs implementation doesn't support mixing of Pedersen generators. So we have to have a range proof per token id in the transaction. (in this version, range_proof_bytes is empty)

This function works with either block version, but after block version 3 is in production, range_proof_bytes can go away.

Reply:
If the upstream bulletproofs supported mixing of Pedersen generators then we wouldn't have to do this and could just have one range proof always.